### ISSUE TITLE
Training on CoinRun / ProcGen

### ISSUE THREAD

**Issue by bmazoure** [astooke/rlpyt#79]
CoinRun (now ProcGen) is quite useful to test for generalization of RL models across multiple levels, and I'm trying to make it run with the library. However, it seems that due to the fact that `env.reset()` are ignored, running the parallel trainer hangs at the `n_itr = self.startup()` stage. What could be a workaround around this issue?

**Comment by astooke** [comment]
Hi!  I haven't used procgen before, taking a quick look at the code now, but maybe you could point to the `env.reset()` being ignored?  I'm not sure where/what is happening with that.  Would definitely be a good environment to get running.

It's possible that everywhere in the sampler where `env.reset()` is called, could subclass the sampler and overwrite that functionality to do whatever is needed for procgen, although this sounds a bit heavy of a solution.

**Comment by bmazoure** [comment]
So the code actually hangs on this line it seems: https://github.com/astooke/rlpyt/blob/master/rlpyt/samplers/buffer.py#L23 
which in turns leads to:
https://github.com/astooke/rlpyt/blob/master/rlpyt/samplers/buffer.py#L60
Any idea why it would happen @astooke ?

**Comment by bmazoure** [comment]
Update: I'm referencing the issue from procgen here.
[https://github.com/openai/procgen/issues/15](https://github.com/openai/procgen/issues/15)

**Comment by christopherhesse** [comment]
@astooke the example that @bmazoure posted does a fork *after* creating the environment, is this what rlpyt does?

If so, that's a bit unreliable since a number of things do not work with fork (such as procgen since it creates worker threads, tensorflow, and pytorch with GPU support).

The safest method is to use spawn like baselines subproc vec env does: https://github.com/openai/baselines/blob/master/baselines/common/vec_env/subproc_vec_env.py#L59

The next best would be to fork *before* creating the process.

**Comment by bmazoure** [comment]
Single-threaded version of procgen now works with the library.

**Comment by christopherhesse** [comment]
It may still be worth making this change to rlpyt to avoid issues with any non-fork-safe environments.

**Comment by astooke** [comment]
OK yes I'm re-opening this to keep it on the to-do list, to sort out how to take advantage of the vec env feature of proc gen.  

I guess the first thing is to establish that it actually does lead to significant speedups/memory savings vs using multiple instances of the single-threaded environment like the current rlpyt setup.

What rlpyt currently does: 
1) Create one instances of the environment to get example observation, env_info, etc.
2) Delete that instance.
3) Fork parallel sampler workers.
4) Inside each worker, create some number of instances of the environment.
5) Inside each worker, at each step, execute a for loop which steps each env instance in sequence.

Any first thoughts as to path of minimal changes to support procgen?

Thanks!

**Comment by christopherhesse** [comment]
It's not necessary to use the vecenv feature of procgen, it's an interface that is only used in baselines.  It may be nice to avoid using fork() regardless though (you can use spawn instead).

**Comment by astooke** [comment]
OK there is some other work getting rlpyt to support spawn (for example follow #99), and it doesn't sound like the vecenv feature of procgen is essential.  Closing this but please reopen if it seems the vectorization in rlpyt is not quite right and would prefer to use that feature of procgen.
